src/tests/test_rng.cpp

```diff
@@ -135,11 +191,209 @@ class HMAC_DRBG_Reseed_Tests : public Test
+         result.test_throws("broken underlying rng but good entropy sources", [&rng_with_broken_rng_and_es] ()
+                  { rng_with_broken_rng_and_es.random_vec(16); });
+
+         mac = Botan::MessageAuthenticationCode::create("HMAC(SHA-256)");
+         Botan::HMAC_DRBG rng_with_rng_and_broken_es(std::move(mac), Test::rng(), broken_entropy_sources, 16384);
+         result.test_throws("good underlying rng but broken entropy sources", [&rng_with_rng_and_broken_es] ()
+                  { rng_with_rng_and_broken_es.random_vec(16); });
+
+         mac = Botan::MessageAuthenticationCode::create("HMAC(SHA-256)");
+         Botan::HMAC_DRBG rng_with_broken_rng_and_broken_es(std::move(mac), broken_entropy_input_rng, broken_entropy_sources, 16384);
+         result.test_throws("underlying rng and entropy sources broken", [&rng_with_broken_rng_and_broken_es] ()
+                  { rng_with_broken_rng_and_broken_es.random_vec(16); });
+
+         return result;
+         }
+
+      Test::Result test_check_nonce()
+         {
+         Test::Result result("HMAC_DRBG Nonce Check");
+
+         auto mac = Botan::MessageAuthenticationCode::create("HMAC(SHA-256)");
+         if(!mac)
+            {
+            result.note_missing("HMAC(SHA-256)");
+            return result;
+            }
+
+         // make sure the nonce has at least 1/2*security_strength bits
+
+         // SHA-256 -> 128 bits security strength
+         for( auto nonce_size : { 0, 4, 15, 16, 17, 32 } )
+            {
+            if(!mac)
+               {
+               mac = Botan::MessageAuthenticationCode::create("HMAC(SHA-256)");
+               }
+
+            Botan::HMAC_DRBG rng(std::move(mac));
+            result.test_eq("not seeded", rng.is_seeded(), false);
+            std::vector<Botan::byte> nonce(nonce_size);
+            rng.initialize_with(nonce.data(), nonce.size());
+
+            if(nonce_size < 16)
+               {
+               result.test_eq("not seeded", rng.is_seeded(), false);
+               result.test_throws("invalid nonce size", [&rng, &nonce] () { rng.random_vec(16); });
+               }
+            else
+               {
+               result.test_eq("is seeded", rng.is_seeded(), true);
+               rng.random_vec(16);
+               }
+            }
+
+         return result;
+         }
+
+      Test::Result test_prediction_resistance()
+         {
+         Test::Result result("HMAC_DRBG Prediction Resistance");
+
+         auto mac = Botan::MessageAuthenticationCode::create("HMAC(SHA-256)");
+         if(!mac)
+            {
+            result.note_missing("HMAC(SHA-256)");
+            return result;
+            }
+
+         // set max_output_before_reseed = 1, forcing a reseed on every request
+         Request_Counting_RNG counting_rng;
+         Botan::HMAC_DRBG rng(std::move(mac), counting_rng, 1);
+
+         rng.random_vec(16);
+         result.test_eq("first request", counting_rng.randomize_count(), size_t(1));
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Convert this cast to the more verbose C++ style. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=cpp%3AS871)